### PR TITLE
Conditional importation

### PR DESCRIPTION
**Pull request by privat**
First quick and dirty implementation for #1579.
Since the loader is due to a full rewrite (cf #1250), some part were implemented as simple as possible.

It is only a technology preview. A lot of thing are missing including the following:

* do not error if the modules in the rules are not present (eg optional modules not installed). Not an issue currently since all the modules are present in the repository.
* show warnings for useless condition to help the programmer
* implement some visibility rules. `public` is assumed even if it makes no sense and cause issues
* understand qualified modules in the conditions

The last issue might be the more limiting currently as only the conditional module understand qualified names. the potential qualified names in the arguments are accepted but just ignored.
Basically `toto` is ignored in `import foo::bar is conditional(toto::tata)`.

### CODE REVIEW

**Comment by xymus** [comment]
+1 !

File: src/loader.nit [hunk 1]
```diff
@@ -272,6 +272,14 @@ redef class ModelBuilder
 	do
 		var path = search_mmodule_by_name(anode, mgroup, name)
 		if path == null then return null # Forward error
+		return load_module_path(path)
+	end
+
+	# Load and process importation of a given ModulePath
+	#
+	# Basically chain `load_module` and `build_module_importation`
```
Comment (Morriar) [inline src/loader.nit]: chains*

File: src/loader.nit [hunk 2]
```diff
@@ -686,6 +694,21 @@ redef class ModelBuilder
 		return mmodule
 	end
 
+	# Resolve the module identification for a given `AModuleName`.
+	#
+	# This method handle qualified names as used in `AModuleName`.
```
Comment (Morriar) [inline src/loader.nit]: handles*

File: src/loader.nit [hunk 3]
```diff
@@ -746,9 +770,99 @@ redef class ModelBuilder
...
+				# Identify the module to automatically import 
+				var suppath = ci.first.as(ModulePath)
+				var sup = load_module_path(suppath)
+				if sup == null then continue
+
+				# Do nothing if already imported
+				if mmodule.in_importation.greaters.has(sup) then continue label
+
+				# Import it
+				self.toolcontext.info("{mmodule} conditionally imports {sup}", 3)
```
Comment (Morriar) [inline src/loader.nit]: So you made a choice regarding the questions in #1579, you should document it on the issue.

File: src/loader.nit [hunk 4]
```diff
@@ -772,6 +886,19 @@ redef class ModelBuilder
 	# All the loaded modules
 	var nmodules = new Array[AModule]
 
+	# Global list of conditional importation rules.
+	#
+	# Each rule is a "Horn clause"-like sequence of modules.
+	# It means that the first module is the module to automatically import.
+	# The remaining modules are the conditions of the rule.
+	#
+	# Each module is either represented by a MModule (is the module is already loaded) or by a ModulePath if the module is not yet loaded.
```
Comment (Morriar) [inline src/loader.nit]: is the module already loaded*

File: src/loader.nit [hunk 5]
```diff
@@ -772,6 +886,19 @@ redef class ModelBuilder
...
+	#
+	# Rules are declared and applied by `build_module_importation`
```
Comment (Morriar) [inline src/loader.nit]: Missing the final dot on this sentence.

File: src/loader.nit [hunk 6]
```diff
@@ -772,6 +886,19 @@ redef class ModelBuilder
...
+	#
+	# TODO (when the loader will be rewrote): use a better representation and move up rules in the model.
```
Comment (Morriar) [inline src/loader.nit]: be rewritten?

**Comment by Morriar** [comment]
A lot of comments on the documentation but +1

File: src/loader.nit (hunk 3 above)
Comment (privat) [inline src/loader.nit]: Could you precise which specific question I should document?

Comment (Morriar) [inline src/loader.nit]: Where to add the conditional module in the module hierarchy and linearization

**Comment by privat** [comment]
reroll&update

* apply comments (spelling)
* handle mixins (-m)
* use it in lib/app
* rely on it in calculator and friendz

File: src/loader.nit (hunk 3 above)
Comment (privat) [inline src/loader.nit]: done

**Comment by Morriar** [comment]
+1

**Comment by xymus** [comment]
Toujours +1
